snapcraft/plugins/colcon.py

```diff
@@ -0,0 +1,687 @@
+      Arguments to pass to cmake projects. Note that any arguments here which match
+      colcon arguments need to be prefixed with a space. This can be done by quoting
+      each argument with a leading space.
+    - colcon-catkin-cmake-args:
+      (list of strings)
+      Arguments to pass to catkin packages. Note that any arguments here which match
+      colcon arguments need to be prefixed with a space. This can be done by quoting
+      each argument with a leading space.
+    - colcon-ament-cmake-args:
+      (list of strings)
+      Arguments to pass to ament_cmake packages. Note that any arguments here which
+      match colcon arguments need to be prefixed with a space. This can be done by
+      quoting each argument with a leading space.
+"""
+
+import contextlib
+import collections
+import os
+import logging
+import re
+import shutil
+import textwrap
+
+import snapcraft
+from snapcraft.plugins import _ros
+from snapcraft.plugins import _python
+from snapcraft import common, file_utils, formatting_utils, repo
+from snapcraft.internal import errors, mangling
+
+logger = logging.getLogger(__name__)
+
+# Map bases to ROS releases
+_ROSDISTRO_TO_BASE_MAP = {"bouncy": "core18", "crystal": "core18"}
+
+# Map bases to Ubuntu releases
+_BASE_TO_UBUNTU_RELEASE_MAP = {"core16": "xenial", "core18": "bionic"}
+
+_SUPPORTED_DEPENDENCY_TYPES = {"apt", "pip"}
+
+_ROS_KEYRING_PATH = os.path.join(snapcraft.internal.common.get_keyringsdir(), "ros.gpg")
+
+
+class ColconInvalidSystemDependencyError(errors.SnapcraftError):
+    fmt = (
+        "Package {dependency!r} isn't a valid system dependency. Did you "
+        "forget to add it to colcon-packages? If not, add the Ubuntu package "
+        "containing it to stage-packages until you can get it into the rosdep "
+        "database."
+    )
+
+    def __init__(self, dependency):
+        super().__init__(dependency=dependency)
+
+
+class ColconUnsupportedDependencyTypeError(errors.SnapcraftError):
+    fmt = (
+        "Package {dependency!r} resolved to an unsupported type of "
+        "dependency: {dependency_type!r}."
+    )
+
+    def __init__(self, dependency_type, dependency):
+        super().__init__(dependency_type=dependency_type, dependency=dependency)
+
+
+class ColconWorkspaceIsRootError(errors.SnapcraftError):
+    fmt = (
+        "colcon-source-space cannot be the root of the colcon workspace; use a "
+        "subdirectory."
+    )
+
+
+class ColconAptDependencyFetchError(errors.SnapcraftError):
+    fmt = "Failed to fetch apt dependencies: {message}"
+
+    def __init__(self, message):
+        super().__init__(message=message)
+
+
+class ColconPackagePathNotFoundError(errors.SnapcraftError):
+    fmt = "Failed to find package path: {path!r}"
+
+    def __init__(self, path):
+        super().__init__(path=path)
+
+
+class ColconPluginBaseError(errors.PluginBaseError):
+    fmt = (
+        "The colcon plugin (used by part {part_name!r}) does not support using base "
```

Comment:
Can we used unified messages here? Either fix all the others (by use of a common exception) or change this one.

Reply:
Well, they serve kind of different purposes at the moment. The common one is just "this plugin doesn't support this base", and that isn't quite the issue here. This one is "that rosdistro is not compatible with the base you selected, try another combination."

I don't see a way to unify those messages, do you?